modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+     */
+    public DistributedProcess(GridKernalContext ctx, DistributedProcessType type,
+        Function<I, IgniteInternalFuture<R>> exec,
+        CI3<UUID, Map<UUID, R>, Map<UUID, Exception>> finish) {
+        this.ctx = ctx;
+        this.type = type;
+
+        log = ctx.log(getClass());
+
+        ctx.discovery().setCustomEventListener(InitMessage.class, (topVer, snd, msg) -> {
+            if (msg.type() != type)
+                return;
+
+            Process p = processes.computeIfAbsent(msg.processId(), id -> new Process(msg.processId()));
+
+            // May be completed in case of double delivering.
+            if (p.initFut.isDone())
+                return;
+
+            ClusterNode crd = coordinator();
+
+            if (crd == null) {
+                p.initFut.onDone();
+
+                onAllServersLeft();
+
+                return;
+            }
+
+            p.crdId = crd.id();
+
+            if (crd.isLocal())
+                initCoordinator(p, topVer);
+
+            IgniteInternalFuture<R> fut = exec.apply((I)msg.request());
+
+            fut.listen(f -> {
+                if (f.error() != null)
+                    p.resFut.onDone(f.error());
+                else
+                    p.resFut.onDone(f.result());
+
+                ClusterNode crdNode = coordinator();
+
+                if (crdNode != null && !ctx.clientNode())
+                    sendSingleSingleMessage(p, crdNode);
+            });
+
+            p.initFut.onDone();
+        });
+
+        ctx.discovery().setCustomEventListener(FullMessage.class, (topVer, snd, msg0) -> {
+            if (msg0.type() != type)
+                return;
+
+            FullMessage<R> msg = (FullMessage<R>)msg0;
+
+            Process p = processes.get(msg.processId());
+
+            if (p == null) {
+                log.warning("Received finish distributed process message for an uninitialized process " +
+                    "(possible cause is message's double delivering) [msg=" +
+                    msg + ']');
+
+                return;
+            }
+
+            finish.apply(p.id,msg.result(), msg.error());
+
+            processes.remove(msg.processId());
+        });
+
+        ctx.io().addMessageListener(GridTopic.TOPIC_DISTRIBUTED_PROCESS, (nodeId, msg0, plc) -> {
+            if (msg0 instanceof SingleNodeMessage && ((SingleNodeMessage)msg0).type() == type) {
+                SingleNodeMessage<R> msg = (SingleNodeMessage<R>)msg0;
+
+                if (msg.type() == type)
+                    onSingleNodeMessageReceived(msg, nodeId);
+            }
+        });
+
+        ctx.event().addDiscoveryEventListener((evt, discoCache) -> {
+            UUID leftNodeId = evt.eventNode().id();
+
+            for (Process p : processes.values()) {
+                p.initFut.listen(fut -> {
+                    boolean crdChanged = F.eq(leftNodeId, p.crdId);
+
+                    if (crdChanged) {
```

Comment:
1. `crdChanged` variable can be omitted.
2. `F.eq` and `localNodeId().equals` - let's use the same technique to check equality.

Reply:
Done.